### PR TITLE
Fix _host_is_ipv6_address for encrypted strings

### LINKED ISSUE #49363: Synchronize fails with encrypted hosts when using delegate_to
##### SUMMARY
Synchronize module fails when using `delegate_to` with an encrypted list of hosts.
In the example we use 3 machines, an `ansible-master` where the playbook is executed and two machines, `machine1` and `machine2`. We have a file on `machine2` at `/home/vagrant/test_file.txt` and we want to copy it on `machine1`. 
`machine1`'s ssh key is an authorized key in `machine2`, so we use synchronize module on `machine2` with `delegate_to: vagrant@{{ hostvars['machine1'].ansible_host }}`.
Our host file, `hosts.yaml` contains the addresses in strings encrypted with ansible-vault.

The execution of the task fails because of a bug in the function `_host_is_ipv6_address`  that checks if the host is an ipv6:
```python
def _host_is_ipv6_address(self, host):
        return ':' in host
```
https://github.com/ansible/ansible/blob/b4ce4e15716b9d6dbb951295097e36675e1deb85/lib/ansible/plugins/action/synchronize.py#L52

The problem here is that when the host is stored as an encrypted string, the type of the variable `host` passed to this function is not a `String` but type `AnsibleVaultEncryptedUnicode`, which is not iterable. 

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION
```
ansible 2.7.3
  config file = /vagrant/ansible/ansible.cfg
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.6 (default, Nov 23 2017, 15:49:48) [GCC 4.8.4]
```

##### CONFIGURATION

```
DEFAULT_HOST_LIST(/vagrant/ansible/ansible.cfg) = [u'/vagrant/ansible/hosts.yaml']
DEFAULT_PRIVATE_KEY_FILE(/vagrant/ansible/ansible.cfg) = /home/vagrant/.ssh/id_rsa
DEFAULT_REMOTE_USER(/vagrant/ansible/ansible.cfg) = vagrant
DEFAULT_VAULT_PASSWORD_FILE(/vagrant/ansible/ansible.cfg) = /vagrant/ansible/vault_pass
HOST_KEY_CHECKING(/vagrant/ansible/ansible.cfg) = False
RETRY_FILES_ENABLED(/vagrant/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

This has been seen in different machines, both Ubuntu based and Amazon Linux.

##### STEPS TO REPRODUCE
Using the vagrant file provided: 
```
vagrant up
vagrant ssh ansible-master
cd /vagrant/ansible
ansible-playbook test_playbook.yml
```
    

test_playbook.yml: 
```yaml
-
    hosts: machine2
    become: yes
    tasks:
        -
            name: Copy the Dataframe from trainer to API machine
            synchronize:
                src: /home/vagrant/test_file.txt
                dest: /home/vagrant/test_file.txt
                mode: pull
            delegate_to: "vagrant@{{ hostvars['machine1'].ansible_host }}"

```

[vagrant_env.zip](https://github.com/ansible/ansible/files/2632861/vagrant_env.zip)
##### EXPECTED RESULTS

The function `_host_is_ipv6_address` shall probably check if the host variable passed is an encrypted string, OR, the calling function shall decrypt it and pass it as a string. 

##### ACTUAL RESULTS

```paste below
TASK [Copy the Dataframe from trainer to API machine] *******************************************************************************************************************************
task path: /vagrant/ansible/test_playbook.yml:6
The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 139, in run
    res = self._execute()
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 603, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/action/synchronize.py", line 343, in run
    src = self._process_remote(_tmp_args, src_host, src, user, inv_port in localhost_ports)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/action/synchronize.py", line 104, in _process_remote
    return self._format_rsync_rsh_target(host, path, user)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/action/synchronize.py", line 66, in _format_rsync_rsh_target
    if self._host_is_ipv6_address(host):
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/action/synchronize.py", line 52, in _host_is_ipv6_address
    return ':' in host
TypeError: argument of type 'AnsibleVaultEncryptedUnicode' is not iterable

fatal: [machine2]: FAILED! => {
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
```

### PR DESCRIPTION
**Pull request by A-Costa**
##### SUMMARY

Adds a check in `_host_is_ipv6_address` for encrypted strings.
Fixes #49363 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`synchronize.py`

### CODE REVIEW

**Review by A-Costa — COMMENTED** [review]
https://github.com/ansible/ansible/issues/49363#issuecomment-443636257

File: lib/ansible/plugins/action/synchronize.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def _get_absolute_path(self, path):
         return path
 
     def _host_is_ipv6_address(self, host):
-        return ':' in host
+        return ':' in to_text(host)
```
Comment (abadger) [inline lib/ansible/plugins/action/synchronize.py:52]: This looks good except that we should use an errors strategy other than replace here.  How about this?
```suggestion
        return ':' in to_text(host, errors='surrogate_or_strict')
```

**Comment by abadger** [comment]
Made a small suggestion.

If you want this fix backported, it will need  a changelog fragment.  you can find examples of those here: https://github.com/ansible/ansible/tree/devel/changelogs/fragments

**Comment by abadger** [comment]
Merged https://github.com/ansible/ansible/pull/49460 which has the same fix.  Thanks for the colaboration between you two to get this fixed :-)

**Pull request closed without merging**